### ISSUE TITLE
EXT4 gives corrupt data when walking files

### ISSUE THREAD

**Issue by jtsylve** [sleuthkit/sleuthkit#767]
I've come across an ext4 image that seems to produce corrupt `TSK_FS_FILE` entries when using `tsk_fs_dir_walk` with the `TSK_FS_DIR_WALK_FLAG_RECURSE` flag.  It seems to be trying to recurse into deleted directories whose inodes have been reallocated.

I am willing to share the image privately with @bcarrier if needed.

**Comment by bcarrier** [comment]
Hey Joe.  I'm not sure we have anyone who can take a look at this right now, but I'll let you know when we do so that we can figure this out.   How are they corrupt?  Bogus names and such?  

We have code in ext2fs_dent.c to do some basic sanity checks on the fields in the data structure:

```
        if ((inode > fs->last_inum) ||  
            (namelen > EXT2FS_MAXNAMLEN) || (namelen == 0) ||   
            (reclen < minreclen) || (reclen % 4) || (idx + reclen > len)) {
```

I just looked at the data structure and the Ext3/4 version does have an extra field 'type' that we are not validating.  We could add a check for that to see if that helps.   I.e verify that type is less than EXT2_DE_MAX.

**Comment by jtsylve** [comment]
Right, it's total garbage data. I noticed because I was getting invalid utf-8 characters when trying to insert the names into a database.  In my case it appears that the parent inode of the invalid nodes was reallocated, but there are already checks for that in the tsk_fs_dir_walk function, so I don't know what's going on. I suppose it's not being marked as reallocated properly, although fls seems to show it as such.
